### PR TITLE
Implement type immunity removal moves

### PR DESCRIPTION
**Pull request by JakubHanko**
The scope of this PR is to implement the following moves:

1. Odor Sleuth
2. Miracle Eye
3. Foresight

The goal is to make the evasiveness reset and type removal work.

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -1827,7 +1827,7 @@ export class ResetStatsAttr extends MoveEffectAttr {
     target.scene.queueMessage(getPokemonMessage(target, `'s stat changes\nwere eliminated!`));
 
     return true;
-  }
+  }
```
Comment (DustinLin) [inline src/data/move.ts:1830]: lol

File: src/data/move.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { Moves } from "./enums/moves";
 import { ChargeAnim, MoveChargeAnim, initMoveAnim, loadMoveAnimAssets } from "./battle-anims";
 import { BattleEndPhase, MovePhase, NewBattlePhase, PartyStatusCurePhase, PokemonHealPhase, StatChangePhase, SwitchSummonPhase } from "../phases";
 import { BattleStat, getBattleStatName } from "./battle-stat";
-import { EncoreTag } from "./battler-tags";
+import { BattlerTag, EncoreTag, TypeImmunityIgnoreTag } from "./battler-tags";
```
Comment (DustinLin) [inline src/data/move.ts:5]: do we need to import `BattlerTag` here?

**Comment by DustinLin** [comment]
Included a demo for odor sleuth, checking to make sure everything works, it would be great if the author does something similar for the rest that they implemented

---


https://github.com/pagefaultgames/pokerogue/assets/39450497/2ecfe207-de48-45ff-9c60-5778effc758b

**Comment by JakubHanko** [comment]
@DustinLin Thanks for the demo and the review. I cleaned up the nits.

About the recording - I might do in during the weekend. The other moves follow the exactly same logic as Odor Sleuth though. If one works, all should.

**Comment by JakubHanko** [comment]
Adding videos for the other two moves.

https://github.com/pagefaultgames/pokerogue/assets/60473007/d99d7b27-45bc-4037-a1a4-2c70bbcacfa1


https://github.com/pagefaultgames/pokerogue/assets/60473007/e64fc237-56b1-48bf-b6b2-7d3508a80cf1

**Review by bennybroseph — CHANGES_REQUESTED** [review]
Looks like there are conflicts on this currently

Looking to get JSDoc documentation as shown here: [Code Comment FAQ](https://discord.com/channels/1125469663833370665/1239957703820836975)

**Comment by JakubHanko** [comment]
> Looks like there are conflicts on this currently
> 
> Looking to get JSDoc documentation as shown here: [Code Comment FAQ](https://discord.com/channels/1125469663833370665/1239957703820836975)

Added the docs, is it better now?

File: src/data/battler-tags.ts [hunk 3]
```diff
@@ -1145,6 +1145,45 @@ export class MagnetRisenTag extends TypeImmuneTag {
   }
 }
 
+/**
+ * Tag that makes the target drop all of it type immunities.
+ * 
+ * Applied by moves: {@linkcode Moves.ODOR_SLEUTH},
+ * {@linkcode Moves.MIRACLE_EYE} and {@linkcode Moves.FORESIGHT}. 
+ */
+export class IgnoreTypeImmunityTag extends BattlerTag {
```
Comment (bennybroseph) [inline src/data/battler-tags.ts:1154]: this class still needs to adhere to the guidelines

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/battler-tags.ts (hunk 3 above)
Comment (JakubHanko) [reply src/data/battler-tags.ts:1157]: Cleaned up the class and hopefully made it respect the guidelines.

File: src/data/move.ts [hunk 4]
```diff
@@ -3557,6 +3557,41 @@ export class FaintCountdownAttr extends AddBattlerTagAttr {
...
+   * @param move {@linkcode Move} being used
+   * @param args N/A
+   * @returns true if the function succeeds
+   */
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
+    if (!super.apply(user, target, move, args))
+      return false;
+
+    target.summonData.battleStats[BattleStat.EVA] = 0;
+    target.updateInfo();
```
Comment (bennybroseph) [inline src/data/move.ts:3587]: This is modifying the target's stats which is not what this move does

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/move.ts (hunk 4 above)
Comment (JakubHanko) [reply src/data/move.ts:3587]: Oh I see, it doesn't reset the evasiveness just once, does it? The evasiveness is completely ignored after this move is used.

Comment (JakubHanko) [reply src/data/move.ts:3587]: Will change it later today.

Comment (JakubHanko) [reply src/data/move.ts]: Reworked it to completely ignore the target's evasiveness.

**Comment by bennybroseph** [comment]
Can we get a video of the same scenarios from the original comment to show that evasion no longer gets reset?

**Comment by JakubHanko** [comment]
https://github.com/pagefaultgames/pokerogue/assets/60473007/27895635-c127-4b5c-a3a1-c6ad81c128a1

Here is video of Odor Sleuth. Are the other two videos necessary? They all follow the same logic. Either all work or none.

**Comment by FogMP** [comment]
For the most part, these moves are working correctly, evasion and types are being ignored.

But going over some edge cases, mainly Wonder Guard and Mind's Eye (basically same functionality as Odor/Foresight):

Mind's eye allows hitting through Wonder Guard with a fighting type move:

https://github.com/pagefaultgames/pokerogue/assets/171270136/87187c27-0048-4280-99de-a8903c898393

However, after using foresight i can't hit it anymore:

https://github.com/pagefaultgames/pokerogue/assets/171270136/480cf340-bce5-4006-abd2-d26e02e13e4b

And just in case here's Wonder Guard vs just Foresight:

https://github.com/pagefaultgames/pokerogue/assets/171270136/8ada1340-548e-4f2e-88a5-b4ffe8b9c496

Unsure how this interaction is supposed to go, but in theory both Mind's Eye and Foresight should have the same result vs wonder guard. And why it makes Mind's Eye stop working.

**Comment by JakubHanko** [comment]
Will look into the interaction with Wonder Guard soon-ish.

**Comment by JakubHanko** [comment]
https://github.com/pagefaultgames/pokerogue/assets/60473007/d9bf1c23-e01a-490d-a607-24123c9abd5f

Fixed the interaction. I thought I was being smart with the way I initially implemented it.

**Comment by brain-frog** [comment]
I'm sorry this was left un-reviewed for 3 weeks, if you update the merge conflicts ping me in discord and I can review it

**Comment by DayKev** [comment]
Imo there should be some unit tests as well.

**Comment by flx-sta** [comment]
make sure to address the conflicts and the comments

**Comment by flx-sta** [comment]
Closed in favor of #3379

**Pull request closed without merging**